### ISSUE TITLE
Allow content designers to replicate selected content

### ISSUE THREAD

**Issue by willguv** [localgovdrupal/localgov#693]
As a content designer
I want to replicate some content
So I don't have to recreate it from scratch

Several councils have requested this for Events (eg Essex, Brighton & Hove)

We also think there's a need to do this for Subsite Overview, as pages can be often quite complicated to create and content designers may want to reuse them

See same issue for microsites https://github.com/localgovdrupal/localgov_microsites/issues/519. Microsites is already using the Replicate module, but not Replicate UI

**Comment by willguv** [comment]
I'll try doing this with my new gitpod set up

**Comment by markconroy** [comment]
@willguv let's get a definition of what we want to do with replicate module. Installing it will do nothing, as it's just an API module. You need code then to use that API to achieve what we want.

I'm guessing we want to be able to clone all/some content types, and/or only certain users can use the clone functionality.

**Comment by willguv** [comment]
Aha! I installed it and there was no visible change - that explains it

@dedavidson has done Replicate work for Dumfries. What did you do there please Duncan?

**Comment by markconroy** [comment]
If we get a list of what we want (or even some of what we want), I'll try spend time on this next week.

Maybe a generic approach with a settings form to say

1. What content types should be replicatable?
2. What user roles should be allowed to use replicate?

Then ... it would be up to each council to decide what way the want to use the module?

**Comment by markconroy** [comment]
@willguv try installing the replicate_ui module and that should give you access to click some options for what you want.

https://www.drupal.org/project/replicate_ui

**Comment by dedavidson** [comment]
We used [Entity Clone](https://www.drupal.org/project/entity_clone) on D&G

**Comment by willguv** [comment]
Was there a reason you picked Entity Clone over Replicate @dedavidson? Thanks

**Comment by dedavidson** [comment]
@willguv nothing in particular except Replicate hasn't been updated in over a year and Entity Clone has more than double the sites using it.

**Comment by willguv** [comment]
@dedavidson @markconroy I've just installed Entity Clone and think it might be a better bet

- it largely works out of the box (just needs permissions to be set)
- for the reasons Duncan mentioned about usage and updates

What do you think please?

I'm not a fan of the word "Clone" if I'm honest. We can discuss changing it if we go this route

**Comment by markconroy** [comment]
I've no problem with us using Entity Clone module.

**Comment by willguv** [comment]
Thanks @markconroy let's go for it then!

Ideally Entity Clone should be added to core and enabled by default

Is there a way to change the word 'Clone' on the button to something else?

**Comment by ekes** [comment]
We use replicate in microsites. Is there really a reason to have two modules that do the same thing in LGD? We also have had this discussion on an issue somewhere else I think? Ah! Here https://github.com/localgovdrupal/localgov_events/issues/96 
The reason for replicate is that it can be controlled and configured by the developers to remove complexity, you can configure custom cloning (making sure the correct referenced entities are cloned etc.). Entity clone can't be used programatically. So one works both ways, the other not.

**Comment by willguv** [comment]
OK that's interesting, thanks @ekes. We shouldn't have two modules doing the same thing

What do we need to use Replicate - a list of entities we would like it to be able to clone?

**Comment by andybroomfield** [comment]
I thought I posted on another issue, my question is what criteria would we prefer one over the other? Does paragraphs get cloned correctly with entity clone. Currently I'm building a replicate book feature for a BHCC section and possible expansion to publications with Replicate, is there an API to do that with entitiy clone as I wouldn't want to have duplicate modules.

**Comment by ekes** [comment]
> is there an API to do that with entitiy clone

No. It's UI, or specifically, form driven. Which is why we used replicate in Microsites. Where as with replicate, it is API first, but you can install, or build, a UI for it.

> What do we need to use Replicate - a list of entities we would like it to be able to clone?

Yes, assuming the replicate UI module isn't up to the job... it's not something I've used.

What seems best is: Make a list of entities you want a 'clone' (or whatever) tab on; and decide by default which entity reference fields you also want duplicated, and which want to reference the existing entity. Generally I think this will be most 'text' paragraphs want duplicating, but stuff like media or geo want to reference the same entity. There might be other things for specific fields that help usability, like do you want to replicate the event into a form with the date empty?

**Comment by willguv** [comment]
Andy: Issue with paragraphs and Entity Clone
Ryan: Have to republish paragraphs with Entity Clone
PaulB: Entity Clone issues with Subsites

**Comment by willguv** [comment]
Finn: Write specific use cases

**Comment by andybroomfield** [comment]
BHCC has gone with Replicate and we've writen a module I could share to replicate whole books.

**Comment by willguv** [comment]
Haringey went with Replicate, also used in LGD microsites

Starshot went with Quck Node Clone

**Comment by markconroy** [comment]
What is the first content type to tackle with this? Do councils want to replicate Service Landing Pages or Subsite Overview pages or something else? Let's choose one, get it done, and then use that as the basis for extending it to more content types.

**Comment by willguv** [comment]
Hi Mark, Event has been mentioned - it's very common for an event to have almost all the same details, maybe just a time or date. This obviously needs to work with current version of events and the new work @joachim-n is doing

If you'd sooner another example to start with I'll have a think

**Comment by markconroy** [comment]
Started here: https://github.com/localgovdrupal/localgov_replicate

Working for all content types, but need to figure out better access checks, so only content types that are set to be replicatable show the replicate tab.

---
Thanks to [Big Blue Door](https://www.bigbluedoor.net/) for sponsoring my time to work on this.

**Comment by markconroy** [comment]
![Screenshot 2024-10-08 at 13 32 16](https://github.com/user-attachments/assets/ead5aaed-dd81-49ec-ae65-3d214e7094da)

**Comment by willguv** [comment]
Hi @markconroy what's left on this please? Thanks

**Comment by markconroy** [comment]
There are [2 open issues](https://github.com/localgovdrupal/localgov_replicate/issues) that need to be looked at.

The first one - Use Replicate UI - needs to be figured out first. If we use that, we might not have much need for my custom module.

**Comment by willguv** [comment]
@markconroy let's either pick this up at today's product group (you're welcome to attend but not mandatory or next week's backlog meet

**Comment by finnlewis** [comment]
From some quick local testing, replicate_ui seems to work OK with simple content and with subsites (which have layout paragraphs). Noted some more points on https://github.com/localgovdrupal/localgov_replicate/issues/2

**Comment by markconroy** [comment]
If replicate UI does what we want, I'm happy for us to use that. It will mean one less module for us to manually maintain.

**Comment by willguv** [comment]
Already used by PaulB at Haringey

**Comment by willguv** [comment]
PB 30 Jan 2025 - we agreed to use Replicate UI out of the box, so content teams can see it in action and suggest changes

@markconroy to create a PR

**Comment by willguv** [comment]
@markconroy as you'll know, Essex is using Replicate for Microsites. Steve at Love Essex has reported that its replicating images as well as other content, FYI

**Comment by willguv** [comment]
@markconroy did you manage to create a PR for this? Thanks

**Comment by markconroy** [comment]
@willguv Yes, but the discussion moved on to more backendy talk, so we decided to pass this over to @ekes or @stephen-cox or someone like that.

**Comment by ekes** [comment]
Was the conclusion https://github.com/localgovdrupal/localgov/issues/693#issuecomment-2607206963 Replicate UI?

**Comment by willguv** [comment]
Planning 3rd April, we're using Replicate UI

**Comment by willguv** [comment]
Planning 1st May, we think this involves:

- install Replicate and UI
- switch on and enable Events, Subsites Overview

If councils want more they will ask for it

**Comment by willguv** [comment]
@markconroy also mentioned this https://www.drupal.org/project/content_templates

**Comment by markconroy** [comment]
I think it might be worth also investigating https://www.drupal.org/project/paragraphs_clipboard module to allow copying paragraphs between nodes from one to the other.

**Comment by markconroy** [comment]
Alpha release has now been made for LocalGov Replicate module. 

https://github.com/localgovdrupal/localgov_replicate/releases/tag/1.0.0-alpha1

It's available on packagist so you can get it with `composer require localgovdrupal/localgov_replicate`

**Comment by willguv** [comment]
@willguv to test this in Drupalforge

Consider a sub module for microsites - @tonypaulbarker to link separate issue

**Comment by markconroy** [comment]
Microsites issue: https://www.drupal.org/project/localgov_replicate/issues/3538543

**Comment by willguv** [comment]
At stand up today we agreed to ship LGD and Microsites together

**Comment by markconroy** [comment]
Two items:

1. It now works with microsites - https://www.drupal.org/project/localgov_replicate/issues/3538543 - I created a submodule for the microsites permissions
2. We have a RC release now - https://www.drupal.org/project/localgov_replicate/releases/1.0.0-rc1

Once someone tests the RC release and says "it works", we can then tag this as 1.0.0.

**Comment by ekes** [comment]
Added a 'works for me' on the localgov_replicate_microsites.
